cpp/arcticdb/storage/mongo/mongo_mock_client.cpp

```diff
@@ -0,0 +1,201 @@
+ *
+ * As of the Change Date specified in that file, in accordance with the Business Source License, use of this software will be governed by the Apache License, version 2.0.
+ */
+
+#include <arcticdb/storage/storage_utils.hpp>
+#include <arcticdb/storage/mongo/mongo_client_wrapper.hpp>
+#include <arcticdb/storage/mongo/mongo_mock_client.hpp>
+#include <arcticdb/storage/object_store_utils.hpp>
+#include <bsoncxx/builder/basic/document.hpp>
+#include <mongocxx/exception/operation_exception.hpp>
+#include <mongocxx/exception/bulk_write_exception.hpp>
+#include <mongocxx/exception/query_exception.hpp>
+
+
+namespace arcticdb::storage::mongo {
+
+std::string MockMongoClient::get_failure_trigger(
+        const std::string& key,
+        StorageOperation operation_to_fail,
+        MongoError error_code) {
+    return fmt::format("{}#Failure_{}_{}", key, operation_to_string(operation_to_fail), static_cast<int>(error_code));
+}
+
+template <typename exception_type>
+MongoFailure create_failure(const std::string& message, MongoError error_code) {
+    static_assert(std::is_base_of<mongocxx::operation_exception, exception_type>::value, "exception_type must be a subclass of mongocxx::operation_exception");
+    if(error_code == MongoError::NoAcknowledge) return {std::monostate()};
+
+    bsoncxx::document::value empty_doc_value = bsoncxx::builder::basic::document{}.extract();
+    auto ec = std::error_code((int)error_code, std::generic_category());
+
+    return {exception_type(ec, std::move(empty_doc_value), message)};
+}
+
+MongoFailure get_failure(const std::string& message, StorageOperation operation, MongoError error_code) {
+    switch(operation) {
+        case StorageOperation::READ:
+            return create_failure<mongocxx::query_exception>(message, error_code);
+        case StorageOperation::WRITE:
+            return create_failure<mongocxx::bulk_write_exception>(message, error_code);
+        case StorageOperation::EXISTS:
+            return create_failure<mongocxx::query_exception>(message, error_code);
+        case StorageOperation::DELETE:
+            return create_failure<mongocxx::bulk_write_exception>(message, error_code);
+        case StorageOperation::DELETE_LOCAL:
+            return create_failure<mongocxx::bulk_write_exception>(message, error_code);
+        case StorageOperation::LIST:
+            return create_failure<mongocxx::query_exception>(message, error_code);
+        default:
+            util::raise_rte("Unknown operation used for error trigger");
+    }
+}
+
+std::optional<MongoFailure> MockMongoClient::has_failure_trigger(
+        const mongo_document_key& key,
+        StorageOperation operation) const {
+    auto key_id = get<2>(key);
+    auto failure_string_for_operation = "#Failure_" + operation_to_string(operation) + "_";
+    auto position = key_id.rfind(failure_string_for_operation);
+    if (position == std::string::npos) return std::nullopt;
```

Comment:
avoid one line if

(applies to several pleaces in this PR)

Reply:
Fixed in [3283f32](https://github.com/man-group/ArcticDB/pull/1395/commits/3283f32c0a323b10940abd25d0b876ca8b57984e)

Reply:
You still have,

```
    if (position == std::string::npos)
        return std::nullopt;
```

should be:

```
    if (position == std::string::npos) {
        return std::nullopt;
    }
```

Reply:
I agree . We have https://github.com/man-group/ArcticDB/wiki/Development-Practices but nothing about style. Let's discuss at the retro.